aten/src/ATen/native/cuda/ForeachBinaryOpScalarTensor.cu

```diff
@@ -0,0 +1,147 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/Dispatch.h>
+#include <ATen/native/BinaryOps.h>
+#include <ATen/native/ForeachUtils.h>
+#include <ATen/native/cuda/ForeachFunctors.cuh>
+#include <ATen/native/cuda/ForeachMinMaxFunctors.cuh>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/_foreach_mul_native.h>
+
+#include <ATen/ops/empty_like_native.h>
+#endif
+
+namespace at::native {
+
+template <typename T, template <class> class Op>
+std::vector<Tensor> foreach_binary_op(TensorList tensors, const Tensor& other) {
+  TORCH_CHECK(
+      other.dim() == 0 && other.numel() == 1,
+      "scalar expected to be 0 dim but ",
+      other.dim(),
+      " and has ",
+      other.numel(),
+      " elements.");
+  TORCH_CHECK(
+      tensors[0].device() == other.device(),
+      "scalar expected to be on ",
+      tensors[0].device(),
+      " but ",
+      other.device());
+  std::vector<std::vector<at::Tensor>> tensor_lists;
+  std::vector<at::Tensor> vec_res;
+  vec_res.reserve(tensors.size());
+  for (const auto& t : tensors) {
+    vec_res.emplace_back(at::native::empty_like(t));
+  }
+
+  tensor_lists.emplace_back(tensors.vec());
+  tensor_lists.emplace_back(std::move(vec_res));
+
+  using opmath_t = at::opmath_type<T>;
+  multi_tensor_apply<2>(
+      tensor_lists,
+      BinaryOpScalarTensorFunctor<
+          T,
+          /* depth */ 2,
+          /* r_args_depth */ 1,
+          /* res_arg_index */ 1>(),
+      Op<opmath_t>(),
+      other.data_ptr<T>());
+  return tensor_lists[1];
+}
+
+template <typename T, template <class> class Op>
+void foreach_binary_op_(TensorList tensors, const Tensor& other) {
+  TORCH_CHECK(
+      other.dim() == 0 && other.numel() == 1,
+      "other expected to be 0 dim but ",
+      other.dim(),
+      " and has ",
+      other.numel(),
+      " elements.");
+  TORCH_CHECK(
+      tensors[0].device() == other.device(),
+      "other expected to be on ",
+      tensors[0].device(),
+      " but ",
+      other.device());
+  std::vector<std::vector<at::Tensor>> tensor_lists;
+  tensor_lists.emplace_back(tensors.vec());
+
+  using opmath_t = at::opmath_type<T>;
+  multi_tensor_apply<1>(
+      tensor_lists,
+      BinaryOpScalarTensorFunctor<
+          T,
+          /* depth */ 1,
+          /* r_args_depth */ 1,
+          /* res_arg_index */ 0>(),
+      Op<opmath_t>(),
+      other.data_ptr<T>());
+  increment_version(tensors);
+}
+
+// TODO(crcrpar): Nest dispatch by looking up `scalar.scalar_type` for better
+// coverage?
+#define FOREACH_BINARY_OP_SCALAR_TENSOR(FUNCTION, NAME, OP, DIVISION_OP) \
+  void foreach_tensor_##NAME##_tensor_kernel_cuda_(                      \
+      TensorList tensors, const Tensor& scalar) {                        \
+    check_foreach_api_restrictions(tensors);                             \
```

Comment:
we need this in ones defined in ForeachOpsKernels.cpp since CPU TensorList inputs don't come to here. would this answer your question?

Reply:
Yea, I realized that. However, could we postpone this check to after we run the slow path (since that will check this anyway)? Or is the runtime for check_foreach_api_restrictions negligible? This also doesn't need to happen in this PR, I'm just curious why the precedent was set as such.

Reply:
Hmm, I'm not sure about the benefit of it because we call this check in the fast path at some point and what it does is basically comparing the size of tensorlists and scalarlists.
I once thought of merging these two into one for fast path while keeping the first check for the slow path.